src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRoslynCompilationService.cs

```diff
@@ -196,7 +198,7 @@ private CSharpCompilation Rewrite(CSharpCompilation compilation)
                 rewrittenTrees.Add(rewrittenTree);
             }
 
-            return compilation.RemoveAllSyntaxTrees().AddSyntaxTrees(rewrittenTrees);
+            compilationContext.Compilation = compilation.RemoveAllSyntaxTrees().AddSyntaxTrees(rewrittenTrees);
```

Comment:
I believe there is also a `ReplaceSyntaxTrees` which should save a step #ImmutabilityProblems